### PR TITLE
cody-gateway: instrument notifications

### PR DESCRIPTION
**Pull request by bobheadxi**
Closes https://github.com/sourcegraph/sourcegraph/issues/53284

Poked at this a bit during company meeting 😛 

## Test plan

CI, `sg start dotcom`

image

### CODE REVIEW

File: enterprise/cmd/cody-gateway/internal/notify/rate_limit.go [hunk 1]
```diff
@@ -9,126 +9,169 @@ import (
...
+	}()
+
+	if slackWebhookURL == "" {
+		logger.Debug("new usage alert",
+			log.Object("actor",
+				log.String("id", actorID),
+				log.String("source", string(actorSource)),
+			),
+			log.String("feature", string(feature)),
+			log.Int("usagePercentage", int(usageRatio*100)),
```
Comment (unknwon) [inline enterprise/cmd/cody-gateway/internal/notify/rate_limit.go:135]: We can get rid of `usageRatio` from the argument list if we just use the `usagePercentage` here?

File: enterprise/cmd/cody-gateway/internal/limiter/limiter.go [hunk 2]
```diff
@@ -165,9 +170,13 @@ func (l StaticLimiter) TryAcquire(ctx context.Context) (_ func(context.Context,
...
-			go l.RateLimitAlerter(float32(currentUsage+usage)/float32(l.Limit), alertTTL)
+			go l.RateLimitAlerter(backgroundContextWithSpan(ctx), float32(currentUsage+usage)/float32(l.Limit), alertTTL)
 		}
 
 		return nil
 	}, nil
 }
+
+func backgroundContextWithSpan(ctx context.Context) context.Context {
+	return trace.ContextWithSpan(context.Background(), trace.SpanFromContext(ctx))
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: If `ctx` has a trace, that trace is extracted using `trace.SpanFromContext` and then injected into a blank background context using `trace.ContextWithSpan` - does that make sense?

File: enterprise/cmd/cody-gateway/internal/notify/rate_limit.go (hunk 1 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/notify/rate_limit.go:135]: Ah good catch, simplified!

File: enterprise/cmd/cody-gateway/internal/limiter/limiter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: Hmm, yes and no? The followup question would be if what we do is extract from the context, then put back into another context, why can't we just pass along the original context?

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: The original context may have cancellation that we don't want to respect, notably during notifications we don't want the notification to be interrupted when the request context is cancelled for whatever reason

[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: Ah I see, if in the cases of we _do_ want to respect cancellation, we just do `tracer.Start(ctx, ...)` and `trace.SpanFromContext(ctx)` accordingly?

[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: Yep! Extracting traces is an edge case for cancellation

[the remark this answers is not in the document]
Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/limiter/limiter.go:181]: Gotcha, thank you!

**Pull request merged**